### PR TITLE
Add doctest tox environment for Python 3.4 and fix doctests

### PR DESCRIPTION
**Pull request by henri-hulski**
Two tests didn't pass.
In `reify.py` I added parentheses to the print statement.

The second failure is in testing.rst:
```
   >>> response
   <200 OK text/plain body='Hello world!'>
```
Python 2 returns here a string as in the example, but Python 3 returns a binary for the body:
```
   >>> response
   <200 OK text/plain body=b'Hello world!'>
```
@taschini Do you've an idea, how to fix this?

### CODE REVIEW

**Comment by taschini** [comment]
It's not so easy I'm afraid. Doctests rely on the equality of repr to validate the tests, and reprs are among the topics where the difference between Python 2 and Python 3 is most obvious.

Moreover, whereas Py.test provides some options to alleviate those differences, Sphinx provides none.

There are ways to recast the examples in the docs to avoid them, but this usually comes at the expense of readability.

Therefore, I think it would make sense if we first asked ourselves whether we really want to verify the examples with both Python 2 and Python 3.

**Comment by henri-hulski** [comment]
Be aware that running the doc tests as described in doc/developing.rst will fail when the developer uses python 3 and our doctests are only compatible with python 2 and vice versa. 

At the moment there is only one test failing.

**Comment by henri-hulski** [comment]
Maybe we should just remove this test.

**Comment by taschini** [comment]
Of course one can add ``# doctest: +SKIP`` to avoid testing the example, but can you please make first the point of why you want to run the doctests with both Python 2.7 and Python 3.4?

Some projects, e.g., [Aiohttp](http://aiohttp.readthedocs.io/en/stable/) state loud and clear on the first page that, despite supporting both Python 3.4 and Python 3.5, the examples in the docs will follow Python 3.5 syntax. We could just add a similar disclaimer to ``doc/developing.rst``.

In fact, it is quite common for projects that support multiple versions of Python to bless one of them and use that for the doctests. A few examples:
- [repoze.lru](https://github.com/repoze/repoze.lru/blob/master/tox.ini): python2.6
- [cryptography](https://github.com/pyca/cryptography/blob/master/tox.ini): python2.7
- [Pyramid](https://github.com/Pylons/pyramid/blob/master/tox.ini): python3.5
- [transaction](https://github.com/zopefoundation/transaction/blob/master/tox.ini): python2.7

There are of course some projects that run doctests with multiple versions, e.g., [Tornado](https://github.com/tornadoweb/tornado/blob/master/tox.ini), but it is a choice that needs to be motivated.

**Comment by henri-hulski** [comment]
Still have to update `CHANGES.txt`, but I prefer to wait until PR #449 is merged to avoid merge conflicts.

**Comment by henri-hulski** [comment]
Now doc test ignores the failed test.

@taschini I think it's natural that if we support Python 2 and 3 we also support doc tests for them. Specially if we're actively promoting them in the documentation and testing some functionality like YAML configuration only in the doc tests. 

I don't know why other projects don't do that, but I think there should be a reason to not support both, not the other way around.

**Comment by taschini** [comment]
Let me go back one step. People use doctests for two reasons:

1. As a way to write literate tests: you write your tests in the style of literate programming, as a document that you can read. See for instance [zest.releaser](https://github.com/zestsoftware/zest.releaser/tree/master/zest/releaser/tests). The doctests are then part of your test suite, even to the point that, as in the case of zest.releaser, they make up the whole test suite.

2. As a way to automatically proof-check the code examples in the docs, as the projects mentioned in my last comment all do.

In Morepath we are using the doctests with the second idea in mind: we have a test suite and we have docs, and we want to limit the risk that the examples in the docs are wrong.

Some code examples can be valid in one version of Python and not in a different version of Python. If the two versions of Python in question are 2.7 and 3.x you, anything producing a string is potentially troublesome.  There are ways of minimizing the situation where this difference is apparent -- Denis mentioned a few in a different thread -- but they have the adverse effect of reducing the readability of examples.

An other possible solution is simply to skip testing those examples, as you are willing to do here and as Tornado [often does](https://raw.githubusercontent.com/tornadoweb/tornado/master/docs/guide/coroutines.rst).  This unfortunately neither addresses the fact that an example is valid in one version but not in the other, and furthermore removes the proof check that at least it is in fact valid for the version for which it is supposed to be valid.

Yet another possible solution is to have different examples (if you are on Python 3 to this, if you are on Python 2 to that), but this does not enjoy built-in support by Sphinx, most of the times it's hardly useful, and typically it is detrimental to the narrative flow.

The final alternative is to accept that the examples will be valid in only one version of Python and use that to run the doctests. This is what Pyramid and the others do.

Now, would running the doctests with both Python 2.7 and 3.4 help improving the correctness of the examples in the code? No. In fact, despite having still quite few tested example in the docs, the first immediate consequence of this change would be to skip an example that was previously tested.

Summing up:

1. We run doctests to improve the correctness of the examples in the docs, not to test functionality.
2. Running doctests with both Python 2 and 3 does not help improve that correctness, and in fact reduces the test coverage of the examples.

**Comment by henri-hulski** [comment]
What are you actually testing with the failed test, which I want to skip? 

```python 
   >>> response
   <200 OK text/plain body='Hello world!'>
```

I don't see any value for this test as the output is rather obvious. 

On the other other hand, by providing the PY3 doctests I found some missing parentheses, which where easy to fix and which also was inconsistent with the examples nearby.

Also when writing the documentation for init_settings I found some inconsistencies and problems with Python 3 by providing the doc tests for Python 3.

So it's generally a good thing to have doc tests for both, Python 2 and 3.

**Comment by taschini** [comment]
> What are you actually testing with the failed test, which I want to skip?

You see, *it's not a test*. It's an *example* and the purpose of the doctest is to proof check the example, not to test anything.  Let me repeat that again: we write the docs, the docs contain examples, we proof-check the examples, and to do that automatically we use doctests.

As trivial as the example might be, it's hardly unique in the kind of problems it presents. The moment we expand doctest coverage from the current four files (and the current coverage for those four is still incomplete) to the rest, we're bound to find again this kind of situation (think for instance at the JSON examples in ``doc/view.rst``).

Now, the cases you mention are code fragments that happen to be valid Python 2 but not Python 3: _there is nothing wrong about that_.  Some of them can be rewritten so that they are valid in both Python 2 and Python 3, some of them cannot.  If you feel that it would be better to adopt Python 3 for the examples, I'm fine with that. However, adopting both Python 2 and Python 3 does not bring any advantage and incurs in definite costs (you'll loose coverage) and that's why using one Python version is a common choice among projects supporting multiple versions of Python.

So, in case it wasn't clear, I'm -1 on this one. I'll let the others weigh in.

**Comment by href** [comment]
I agree with @taschini here. We want to verify that the examples in the docs actually work. And those examples ultimately don't gain anything from supporting both Python 2 and Python 3. If we support both we will end up with example code that needs to be aware of differences between the two. That's outside the scope of this project imho.

So this sounds very reasonable to me:

> Some projects, e.g., Aiohttp state loud and clear on the first page that, despite supporting both Python 3.4 and Python 3.5, the examples in the docs will follow Python 3.5 syntax. We could just add a similar disclaimer to doc/developing.rst.

I'm +1 on only supporting the latest python release in our docs. People who stick with Python 2.7 or Pypy are more likely to know about the differences to 3.5 and can apply that knowledge. Beginners will probably start with 3.x at this point, so we should optimise for them.

It should be no secret though that I already switched most of my stack to 3.x, so I'm biased ;)

**Comment by henri-hulski** [comment]
@faassen 
What is your opinion here? 

If you're all agree I'll move the doc tests from Python 2.7 to 3.5 and leave a note in testing.rst.

**Comment by taschini** [comment]
> If you're all agree I'll move the doc tests from Python 2.7 to 3.5 and leave a note in testing.rst.

I agree!
